src/correction.cc

```diff
@@ -45,25 +96,186 @@ Formula::Formula(const rapidjson::Value& json) :
     type_ = ParserType::numexpr;
     throw std::runtime_error("numexpr formula parser is not yet supported");
   }
-  else { throw std::runtime_error("Unrecognized formula parser type"); } 
+  else { throw std::runtime_error("Unrecognized formula parser type"); }
 
   for (const auto& item : json["parameters"].GetArray()) {
-    parameterIdx_.push_back(item.GetInt());
+    variableIdx_.push_back(item.GetInt());
   }
+
+  if ( eager_compilation ) build_ast();
 }
 
-double Formula::evaluate(const std::vector<Variable>& inputs, const std::vector<Variable::Type>& values) const {
-  if ( ! evaluator_ ) {
-    // TODO: thread-safety: should we acquire a lock when building?
-    evaluator_ = std::make_unique<TFormula>("formula", expression_.c_str(), false);
-    if ( evaluator_->Compile() != 0 ) {
-      throw std::runtime_error("Failed to compile expression " + expression_ + " into TFormula");
+void Formula::build_ast() const {
+  std::shared_ptr<peg::Ast> peg_ast;
+  {
+    const std::lock_guard<std::mutex> lock(parsers_mutex_);
+    auto& parser = parsers_[type_];
+    if ( ! parser ) {
+      if ( type_ == ParserType::TFormula ) parser.load_grammar(TFormula_grammar_v1);
+      parser.enable_ast();
+      parser.enable_packrat_parsing();
+    }
+    int pos;
+    std::string msg;
+    parser.log = [&](size_t ln, size_t col, const std::string &themsg) {
+      pos = col;
+      msg = themsg;
+    };
+    if ( ! parser.parse(expression_, peg_ast) ) {
+      throw std::runtime_error(
+        "Failed to parse Formula expression at position " + std::to_string(pos) + ":\n"
+        + expression_ + "\n"
+        + std::string(pos, ' ') + "^\n"
+        + msg
+      );
     }
+    peg_ast = parser.optimize_ast(peg_ast);
+  }
+  ast_ = std::make_unique<Ast>(translate_ast(*peg_ast));
```

Comment:
Do you know or sure that using peg_ast at this point is thread safe?

Reply:
translate_ast takes a const reference, dunno if that's a guarantee. At least it appears to be a data container only.

Reply:
Looking at the peg code, I'm having a very hard time being able to prove it generates stuff that is reentrant. Lots of holding pointers, having mutating methods (e.g. the visitor pattern they use allows mutating the items in the node) as well as the use of `mutable`.

Reply:
Ok, I'll move it inside the locked scope. This was pretty clear premature optimization